### DISCUSSION TITLE
Multivariate series with of different lengths

### DISCUSSION THREAD

**Discussion by elisevans** [TDAmeritrade/stumpy#905] [Help: Coding & Implementations]
If I have multiple time series but not all of equal length (e.g. some have 100 target values, some have 98 target values) is there a way to still use mstump, or should I iterate over each series and use stump?

I tried doing the former by inserting np.nan values but this leads to inf values in the matrix profile. Not sure if there is some way around this.

**Comment by seanlaw** [comment]
@elisevans Thank you for your question and welcome to the STUMPY community. 

> If I have multiple time series but not all of equal length (e.g. some have 100 target values, some have 98 target values) is there a way to still use mstump, or should I iterate over each series and use stump?

So, for multi-dimensional cases, it is important to emphasize that the multi-dimensional matrix profile is NOT simply 1-dimensional matrix profiles stacked one on top of the other. More information can be found [here](https://stumpy.readthedocs.io/en/latest/Tutorial_Multidimensional_Motif_Discovery.html). `mstump` assumes that ALL of your time series are aligned (i.e., all data points are captured at the same time points and frequency) and so having different lengths implies that this condition is broken and you must decide how to align the missing time points. 

> I tried doing the former by inserting np.nan values but this leads to inf values in the matrix profile.

This is to be expected since it is impossible to compute distances between a subsequence with missing values. Fundamentally, there is nothing wrong with `inf` values if they are related to subsequences with `np.nan` values. If indeed your time series are unrelated then you should just use `stump` individually over each time series as there is no performance gain otherwise.

Can you please describe what it is that you are hoping to accomplish? This added context might allow us to provide further guidance.

**Reply by elisevans** [reply]
Thank you for the quick response!

I have multiple time series of the weekly sales of similar products, and my use case is exploring whether out-of-stock instances can be detected via anomaly detection by the matrix profile.

Whilst they are all captured at the same frequency (weekly) some products started being sold at a later time than others, so I suppose that violates the alignment requirement for mstump.

**Reply by NimaSarajpoor** [reply]
Not sure how much the following notes/questions can be useful in your case but I am going to share with you anyway. Thinking about them might help you :) 

1) what is the value in your time series? Is it the number of products in stock? Or is it the number of items being sold? I was wondering if answering these kind of question can help you decide how to fill the nan value.

2) why not concatenate them and make one time series?

3) what does dropping nan mean in your case? Note that matrix profile does NOT need the time stamps to have data recorded in a fix frequency. However, that does not necessary mean that you can drop nan. It depends on your problem and what dropping nan means.

4) If you think there should be a common similar pattern/motif (not anomaly) among those time series, you may use ostinato first to find common pattern. However similar patterns across your several timeseries may not start at the same time stamp. Now, another question: Does that matter? In other words٫ should all similar patterns start at the same time stamp? If yes, you may NOT use ostinato.


My suggestion:
Read STUMPY tutorial/examples on mstump and ostinato.(I assume you are already familiar with stump, single dimensional matrix profile)

Then think about your problem.

Then see if you can find a connection.
